[clang-repl] Emit const variables only once

hahnjo opened the pull request:
Disable internal linkage for const variables if IncrementalExtensions are enabled. Otherwise the variables are emitted multiple times, with multiple constructions at unique memory locations, during every PTU.

vgvassilev reviewed (approved):
That looks good to me. Can we backport that to the llvm17 release branch? I know at least one downstream project that jumps through hoops to support this.

AaronBallman commented:
> That looks good to me. Can we backport that to the llvm17 release branch? I know at least one downstream project that jumps through hoops to support this.

I'd say we should not backport to 17.0.0; we need to get the final rc out the door and this isn't fixing a critical bug or regression. However, it may be reasonable to consider for 17.0.1+

hahnjo committed:
[clang-repl] Emit const variables only once

Disable internal linkage for const variables if IncrementalExtensions
are enabled. Otherwise the variables are emitted multiple times, with
multiple constructions at unique memory locations, during every PTU.

hahnjo commented:
The original patch worked for `clang-repl` but results in strong linkage which I found to cause problems with modules downstream in ROOT. Instead the latest push moves the special case one level higher to `basicGVALinkageForVariable` and returns `GVA_DiscardableODR` which fixes that problem as well.

iains reviewed (commented):
(still getting used to the revised system)
I have no comments on this patch.

hahnjo commented:
ping @vgvassilev

vgvassilev commented on clang/test/Interpreter/const.cpp:
```diff
@@ -0,0 +1,29 @@
+// RUN: cat %s | clang-repl -Xcc -O2 | FileCheck %s
+
+extern "C" int printf(const char*, ...);
+
```
Can you add a case for a file scope constant? We also should consider enabling that for C.

hahnjo replied:
`const A a(1);` is a file-scope constant, no? We don't need it for C because the special case in `LinkageComputer::getLVForNamespaceScopeDecl` only applies to C++ (my understanding is that `const` variables always have an identity in C).

vgvassilev replied:
> const A a(1); is a file-scope constant, no?

Yes, missed that.

vgvassilev reviewed (approved):
LGTM. Let's move forward here and rely on a further post-commit review if necessary.

hahnjo commented:
Note that this currently doesn't seem to work on Windows: https://github.com/llvm/llvm-project/issues/68092
